### PR TITLE
add fix for metricsProperties when HasPrometheusConfigFile is true.

### PR DESCRIPTION
**Pull request by jinxingwang**
Fix the bug where Monitor.Prometheus.ConfigFile is set. the configMap which contains spark metrics.properties & prometheus config file will not be created & mounted, so the metricsProperties will be useless. 
Add a new property metricsPropertiesFile, this will overwrite the value `spark.metrics.conf` in spark.properties. give a better way to specify spark metrics configuration when it is large.

### CODE REVIEW

**Comment by jinxingwang** [comment]
@liyinan926 Please review.
When I am running the build commands such as:
go generate
controller-gen ...
make build-api-docs

There are auto-generate code for previous PRs such as lifeCycle, TerminationGracePeriod
Are those not included intentionally?

**Comment by liyinan926** [comment]
The authors of the previous PRs must have forgotten to run these. Please include them in this PR.

File: pkg/apis/sparkoperator.k8s.io/v1beta2/types.go [hunk 1]
```diff
@@ -557,6 +557,10 @@ type MonitoringSpec struct {
 	// +optional
 	// If not specified, the content in spark-docker/conf/metrics.properties will be used.
 	MetricsProperties *string `json:"metricsProperties,omitempty"`
+	// MetricsPropertiesFile is the address of file metrics.properties for configuring the Spark metric system.
```
Comment (liyinan926) [inline pkg/apis/sparkoperator.k8s.io/v1beta2/types.go]: Suggest change `address` to `the container local path`.

File: pkg/controller/sparkapplication/monitoring_config.go [hunk 2]
```diff
@@ -44,14 +44,8 @@ func configPrometheusMonitoring(app *v1beta2.SparkApplication, kubeClient client
...
 
-	var javaOption string
-	if app.HasPrometheusConfigFile() {
-		configFile := *app.Spec.Monitoring.Prometheus.ConfigFile
-		glog.V(2).Infof("Overriding the default Prometheus configuration with config file %s in the Spark image.", configFile)
-		javaOption = fmt.Sprintf("-javaagent:%s=%d:%s", app.Spec.Monitoring.Prometheus.JmxExporterJar,
-			port, configFile)
-	} else {
-		glog.V(2).Infof("Using the default Prometheus configuration.")
+	if !app.HasMetricsPropertiesFile() || !app.HasPrometheusConfigFile() {
```
Comment (liyinan926) [inline pkg/controller/sparkapplication/monitoring_config.go]: Can we refactor the if-else block like the following to make it easier to read and understand?

```
var javaOption string

if app.HasMetricsPropertiesFile() && app.HasPrometheusConfigFile() {
    glog.V(2).Infof("Loading local metrics and Prometheus configuration files")
    ...
    javaOptions = ...
} else {
    glog.V(2).Infof("Creating a ConfigMap for metrics and/or Prometheus configurations")
    ...
    javaOptions = ...
}
```

File: pkg/controller/sparkapplication/monitoring_config_test.go [hunk 3]
```diff
@@ -50,20 +51,34 @@ func TestConfigPrometheusMonitoring(t *testing.T) {
...
+		if test.app.Spec.Monitoring.Prometheus.ConfigFile == nil &&
+			test.app.Spec.Monitoring.MetricsPropertiesFile == nil &&
+			len(configMap.Data) != 2 {
 			t.Errorf("expected %d data items got %d", 2, len(configMap.Data))
 		}
 
-		if configMap.Data[metricsPropertiesKey] != test.metricsProperties {
+		if test.app.Spec.Monitoring.Prometheus.ConfigFile != nil &&
+			test.app.Spec.Monitoring.MetricsPropertiesFile == nil &&
+			len(configMap.Data) > 1 {
```
Comment (liyinan926) [inline pkg/controller/sparkapplication/monitoring_config_test.go]: `len(configMap.Data) != 1`

File: pkg/controller/sparkapplication/monitoring_config_test.go [hunk 4]
```diff
@@ -50,20 +51,34 @@ func TestConfigPrometheusMonitoring(t *testing.T) {
...
+			t.Errorf("expected %d data items got %d", 1, len(configMap.Data))
+		}
+
+		if test.app.Spec.Monitoring.Prometheus.ConfigFile == nil &&
+			test.app.Spec.Monitoring.MetricsPropertiesFile != nil &&
+			len(configMap.Data) > 1 {
```
Comment (liyinan926) [inline pkg/controller/sparkapplication/monitoring_config_test.go]: `len(configMap.Data) != 1`

**Comment by jinxingwang** [comment]
@liyinan926 Please review again, thanks

File: pkg/controller/sparkapplication/monitoring_config.go [hunk 5]
```diff
@@ -44,8 +44,10 @@ func configPrometheusMonitoring(app *v1beta2.SparkApplication, kubeClient client
 		port = *app.Spec.Monitoring.Prometheus.Port
 	}
 
-	if !app.HasMetricsPropertiesFile() || !app.HasPrometheusConfigFile() {
-		glog.V(2).Infof("Loading Prometheus configuration.")
+	if app.HasMetricsPropertiesFile() && app.HasPrometheusConfigFile() {
+		glog.V(2).Infof("Loading local Spark metrics and Prometheus configuration files, not creating configMap with default properties.")
```
[the remark this answers is not in the document]
Comment (jinxingwang) [reply pkg/controller/sparkapplication/monitoring_config.go]: Got it. now it should look better.

[the remark this answers is not in the document]
Comment (jinxingwang) [reply pkg/controller/sparkapplication/monitoring_config.go]: I think I will set the logic back to the original. with clarify comments.
if app.HasMetricsPropertiesFile() && app.HasPrometheusConfigFile() {
} else {
}
not really cover all the cases. to cover all the case:
if app.HasMetricsPropertiesFile() && app.HasPrometheusConfigFile() {
Overwtie defaul spark metrics file & defaul prometheus config file
} else if app.HasMetricsPropertiesFile() {
Overwtie defaul spark metrics file & create CM with default prometheus config file
} else if app.HasPrometheusConfigFile(){
Overwtie defaul prometheus config file & create CM with default spark metrics file 
} else {
create CM with default spark metrics file & defaul prometheus config file
}

[the remark this answers is not in the document]
Comment (liyinan926) [reply pkg/controller/sparkapplication/monitoring_config.go]: I think you can combine the three `else` cases into one, as you already handle the metrics config file and Prometheus config file individually in `buildPrometheusConfigMap`. So basically there are cases: 1) no metrics config file nor Prometheus config file are specified, and 2) at least one of the two config files is not specified, and the ConfigMap needs to be created.

[the remark this answers is not in the document]
Comment (jinxingwang) [reply pkg/controller/sparkapplication/monitoring_config.go]: If I do that. there will be places for duplicate code blocks. As a example here:

if app.HasMetricsPropertiesFile() && app.HasPrometheusConfigFile() {
        // using PrometheusConfigFile
	javaOption = ...
        // using MetricsPropertiesFile
	app.Spec.SparkConf["spark.metrics.conf"] = ...
} else {
        // Create configMap
	configMap := buildPrometheusConfigMap(app, configMapName)
  
	if app.HasPrometheusConfigFile() {
                // using PrometheusConfigFile
		javaOption = ...
	} else {
                 // using default PrometheusConfigFile
		javaOption = ...
	}
        // using MetricsPropertiesFile
	if app.HasMetricsPropertiesFile() {
		app.Spec.SparkConf["spark.metrics.conf"] = ...
	}

}

File: pkg/controller/sparkapplication/monitoring_config.go [hunk 6]
```diff
@@ -44,14 +44,10 @@ func configPrometheusMonitoring(app *v1beta2.SparkApplication, kubeClient client
...
-		configFile := *app.Spec.Monitoring.Prometheus.ConfigFile
-		glog.V(2).Infof("Overriding the default Prometheus configuration with config file %s in the Spark image.", configFile)
-		javaOption = fmt.Sprintf("-javaagent:%s=%d:%s", app.Spec.Monitoring.Prometheus.JmxExporterJar,
-			port, configFile)
-	} else {
-		glog.V(2).Infof("Using the default Prometheus configuration.")
+	// If one of the metricsPropertiesFile & prometheusConfigFile is not set
+	// mounting configMap to using default configuration
+	if !app.HasMetricsPropertiesFile() || !app.HasPrometheusConfigFile() {
+		glog.V(2).Infof("Loading Prometheus configuration.")
```
Comment (liyinan926) [inline pkg/controller/sparkapplication/monitoring_config.go]: Can we change this to `Creating a ConfigMap for metrics and Prometheus configurations`?

File: pkg/controller/sparkapplication/monitoring_config.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (liyinan926) [reply pkg/controller/sparkapplication/monitoring_config.go]: OK, let's just keep the current structure in the PR.

File: pkg/controller/sparkapplication/monitoring_config.go [hunk 7]
```diff
@@ -72,13 +68,23 @@ func configPrometheusMonitoring(app *v1beta2.SparkApplication, kubeClient client
...
+
+	var javaOption string
 
-		javaOption = fmt.Sprintf(
-			"-javaagent:%s=%d:%s/%s",
-			app.Spec.Monitoring.Prometheus.JmxExporterJar,
-			port,
-			config.PrometheusConfigMapMountPath,
-			prometheusConfigKey)
+	glog.V(2).Infof("Setting the default Prometheus configuration.")
```
Comment (liyinan926) [inline pkg/controller/sparkapplication/monitoring_config.go]: Please also remove this log line that is pretty useless.

**Pull request merged**